### DISCUSSION TITLE
Is nonnegative hierarchical time series reconciliation implemented in sktime?

### DISCUSSION THREAD

**Discussion by elenagutiv** [alan-turing-institute/sktime#4645] [Q&A] [answered]
I'm using sktime to perform time series forecasting and hierarchical reconciliation. For the latter I am using the `ReconcilerForecaster` class with the OLS method. My question is: is there a way to obtain **nonnegative** revised forecasts after the reconciliation step using `ReconcilerForecaster`?

I am aware of other implementations of this restriction for hierarchical TS reconciliation, like the library `hierarchicalforecast `([HierarchicalForecast library](https://nixtla.github.io/hierarchicalforecast/methods.html)) where the MinTrace and OptimalCombination reconciliation classes allow to define the restriction `nonnegative = True` for getting always nonnegative revised forecasts. I wanted to know if there's a way to specify this restriction in sktime?

**Comment by fkiraly** [comment, marked as answer]
hm, interesting question. The `Reconciler` or `ReconcilerForecaster` themselves do not have this.

The easiest way to get nonnegative forecasts are using the `FunctionTransformer` with the function `lambda x: max(0, x)` or similar, and using it in a pipeline *after*  the `ReconcilerForecaster`, i.e., `my_reconciler * FunctionTransformer(nonneg_fun)`

This is not exactly the same as you describe, as this is after applying the reconciler, and the nixtla forecaster has this as a parameter of the method.

We're currently in the process of interfacing the nice nixtla libraries under consistent interfaces:
https://github.com/sktime/sktime/issues/4643
so, if you'd like to help by contributing an interface to the hierarchical forecasters (e.g., `NixtlaReconciler`), that would be useful.

A de-novo implementation of course would also be useful, if it can be easily added by using the internal methods of the `sktime` internal `ReconilerForecaster`.

**Reply by fkiraly** [reply]
FYI @yarnabrina, @ciaran-g

**Reply by elenagutiv** [reply]
Thanks for your quick response @fkiraly! As you mentioned, forcing the nonnegative restriction after reconciling would loose the reconciliation property of forecasts of adding up in a manner that is consistent with the aggregation structure.
Thanks a lot for your continuous contributions to the library, I'll keep mysef posted.
